src/orange/univ/RectArrayTracker.hh

```diff
@@ -181,23 +181,22 @@ RectArrayTracker::cross_boundary(LocalState const& state) const
     auto coords = to_coords(state.volume.unchecked_get());
     auto ax_idx = this->find_surface_axis_idx(state.surface.id());
 
+    // Due to surface deduplication, crossing out of rect arrays is not
+    // possible
+    CELER_EXPECT(
+        !(coords[ax_idx] == 0 && state.surface.sense() == Sense::inside)
+        && !(coords[ax_idx] == record_.dims[ax_idx] - 1
+             && state.surface.sense() == Sense::outside));
+
     // Value for incrementing the axial coordinate upon crossing
-    int inc = (state.surface.sense() == Sense::outside) ? -1 : 1;
+    // NOTE: that surface sense is currently the POST crossing value
+    int inc = (state.surface.sense() == Sense::outside) ? 1 : -1;
 
     detail::OnLocalSurface new_surface(state.surface.id(),
-                                       flip_sense(state.surface.sense()));
+                                       state.surface.sense());
```

Comment:
This is just `state.surface` now, and you can replace the use of it below directly with that.